### PR TITLE
feat: Branch mst adjmatrix

### PR DESCRIPTION
**Pull request by junnengsoo**

### CODE REVIEW

File: src/main/java/algorithms/minimumSpanningTree/kruskal/Edge.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package algorithms.minimumSpanningTree.kruskal;
```
Comment (4ndrelim) [inline src/main/java/algorithms/minimumSpanningTree/kruskal/Edge.java:1]: our other files typically place helper class in the same file. For consistency sake, we could shift this to Kruskal.java too. But if there are any concerns do bring it up!

File: src/main/java/algorithms/minimumSpanningTree/kruskal/DisjointSet.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * Implementation of quick-find structure; Turns a list of objects into a data structure that supports union operations
+ */
+public class DisjointSet {
+    private final Map<Node, String> identifier;
+
+    /**
+     * Constructor to initialize Disjoint Set with a known list of listOfNodes.
+     * @param listOfNodes
+     */
+    public DisjointSet(Node[] listOfNodes) {
```
Comment (4ndrelim) [inline src/main/java/algorithms/minimumSpanningTree/kruskal/DisjointSet.java:16]: nice, but we should align our implementation of disjoint set. There already is an existing one which u can either import or copy-paste (maybe this is preferable? to be explicit).

Ok but i did notice a concern as i was looking through. Both urs and my implementation does not handle duplicates well. I will tweak the official DisjointSet implementation of our repo to identify by memory address instead.

File: src/main/java/algorithms/minimumSpanningTree/kruskal/Kruskal.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            Node source = edge.getSource();
+            Node destination = edge.getDestination();
+            if (!ds.find(source, destination)) {
+                mstMatrix[source.getIndex()][destination.getIndex()] = edge.getWeight();
+                mstMatrix[destination.getIndex()][source.getIndex()] = edge.getWeight();
+                ds.union(source, destination);
+            }
+        }
+
+        return mstMatrix;
```
Comment (4ndrelim) [inline src/main/java/algorithms/minimumSpanningTree/kruskal/Kruskal.java:58]: cant rmb if this was what we discussed. But we do this just as a representation to capture the edges right? Do you know of any online helper functions to visualise the mst lol

File: src/main/java/algorithms/minimumSpanningTree/kruskal/Kruskal.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ *  Starting from any source (this will be the first node to be in the MST), pick the lightest outgoing edge, and
+ *  include the node at the other end as part of a set of nodes S. Now repeatedly do the above by picking the lightest
+ *  outgoing edge adjacent to any node in the MST (ensure the other end of the node is not already in the MST).
+ *  Repeat until S contains all nodes in the graph. S is the MST.
+ * Actual implementation:
+ *  No Edge class was implemented. Instead, the weights of the edges are stored in a 2D array adjacency matrix. An
+ *  adjacency list may be used instead
+ *  A Node class is implemented to encapsulate the current minimum weight to reach the node.
+ */
+public class Kruskal {
```
Comment (4ndrelim) [inline src/main/java/algorithms/minimumSpanningTree/kruskal/Kruskal.java:18]: overall logic is neat and clear! thanks!

File: src/main/java/algorithms/minimumSpanningTree/kruskal/Node.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
+package algorithms.minimumSpanningTree.kruskal;
```
Comment (4ndrelim) [inline src/main/java/algorithms/minimumSpanningTree/kruskal/Node.java:1]: yup same for this class also. Maybe can put at the bottom as helper classes

File: src/main/java/algorithms/minimumSpanningTree/prim/Prim.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+ *  outgoing edge adjacent to any node in the MST (ensure the other end of the node is not already in the MST).
+ *  Repeat until S contains all nodes in the graph. S is the MST.
+ * Actual implementation:
+ *  No Edge class was implemented. Instead, the weights of the edges are stored in a 2D array adjacency matrix. An
+ *  adjacency list may be used instead
+ *  A Node class is implemented to encapsulate the current minimum weight to reach the node.
+ */
+public class Prim {
+    public static int[][] getPrimsMST(Node[] nodes, int[][] adjacencyMatrix) {
+        PriorityQueue<Node> pq = new PriorityQueue<>((a, b) -> a.getCurrMinWeight() - b.getCurrMinWeight());
```
Comment (4ndrelim) [inline src/main/java/algorithms/minimumSpanningTree/prim/Prim.java:20]: maybe a small helper comment to remind its a min-heap

File: src/main/java/algorithms/minimumSpanningTree/prim/Prim.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        int[][] mstMatrix = new int[nodes.length][nodes.length]; // MST adjacency matrix
+
+        int[] parent = new int[nodes.length]; // To track the parent node of each node in the MST
```
Comment (4ndrelim) [inline src/main/java/algorithms/minimumSpanningTree/prim/Prim.java:23]: nice. Map<> works too but i think array is even simpler.

File: src/main/java/algorithms/minimumSpanningTree/prim/Prim.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
+package algorithms.minimumSpanningTree.prim;
```
Comment (4ndrelim) [inline src/main/java/algorithms/minimumSpanningTree/prim/Prim.java:1]: i believe the underlying assumption is that the graph is connected right? i think both ur implementation should work fine for unconnected graphs too (not a thorough analysis.. jsut a quick run-through in my head) - that is, your return matrix should capture separate MSTs, if exist. Do double check and lmk if im wrong!

**Review by 4ndrelim — CHANGES_REQUESTED** [review]
Not done yet, ill update on Friday after i tweak Disjoint Set implementation to better handle duplicates and to be more portable!

**Comment by 4ndrelim** [comment]
@junnengsoo can we close this?

**Comment by junnengsoo** [comment]
A separate PR for MST was merged to main.

**Pull request closed without merging**